src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/Grammar/GrammarGenerator.cs

```diff
@@ -0,0 +1,325 @@
+                    if (sorted.Length > 0)
+                        normalizedRules.Add((Normalize(name), sorted.Select(s => s.Text).ToImmutableArray()));
+
+                    // Now proceed in depth-first fashion through the rules the productions of this rule
+                    // reference.  This helps keep related rules of these productions close by.
+                    //
+                    // Note: if we hit a major-section node, we don't recurse in.  This keeps us from
+                    // travelling too far away, and keeps the major sections relatively cohesive.
+                    var references = sorted.SelectMany(t => t.RuleReferences).Where(r => !s_majorRules.Contains(r));
+                    foreach (var reference in references)
+                        AddNormalizedRules(reference);
+                }
+            }
+        }
+
+        private static string Generate(string name, ImmutableArray<string> productions)
+            => Environment.NewLine + Environment.NewLine + name + Environment.NewLine + "  : " +
+               Join(Environment.NewLine + "  | ", productions) + Environment.NewLine + "  ;";
+
+        /// <summary>
+        /// Returns the g4 production string for this rule based on the children it has. Also
+        /// returns all the names of other rules this particular production references.
+        /// </summary>
+        private Production ProcessChildren(List<TreeTypeChild> children, string delim)
+        {
+            var result = children.Select(child => child switch
+            {
+                Choice c => ProcessChildren(c.Children, " | ").Parenthesize(),
+                Sequence s => ProcessChildren(s.Children, " ").Parenthesize(),
+                Field f => GetFieldType(f).WithSuffix(f.IsOptional ? "?" : ""),
+                _ => throw new InvalidOperationException(),
+            }).Where(p => p.Text.Length > 0);
+
+            return new Production(
+                Join(delim, result.Select(t => t.Text)),
+                result.SelectMany(t => t.RuleReferences));
+        }
+
+        private Production GetFieldType(Field field)
+            // 'bool' fields are for the few boolean properties we generate on DirectiveTrivia.
+            // They're not relevant to the grammar, so we just return an empty production here
+            // which will be filtered out by the caller.
+            => field.Type == "bool" ? new Production("") :
+               field.Type == "CSharpSyntaxNode" ? HandleCSharpSyntaxNodeField(field) :
+               field.Type.StartsWith("SeparatedSyntaxList") ? HandleSeparatedSyntaxListField(field) :
+               field.Type.StartsWith("SyntaxList") ? HandleSyntaxListField(field) :
+               field.IsToken ? HandleSyntaxTokenField(field) : RuleReference(field.Type);
+
+        private static Production HandleSyntaxTokenField(Field field)
```

Comment:
nit: I'm still reading through so my mental model may be incomplete, but I'm a little confused by the naming `Handle` vs. `Get` (in `GetFieldType` above) vs. `Process` (`ProcessChildren`). All three kinds of methods return `Production`s, except for `ProcessProductions`.

Reply:
agreed.  let me try to clean this up.

Reply:
I tried to simplify and rename. LMK if hte current model is clearer!